### PR TITLE
Fix duplicate key

### PR DESCRIPTION
**Pull request by woksin**
## Summary

There could occur an issue when a filter would try to write the same event to a stream multiple times. To fix this we catch the duplicate key exception and tries to figure out which events to write.

### Fixed

- Issue where filter tries to write the same event to a stream forever

### CODE REVIEW

File: Source/Events.Store.MongoDB/EventHorizon/EventsToPublicStreamsWriter.cs [hunk 1]
```diff
@@ -36,41 +34,34 @@ public class EventsToPublicStreamsWriter : IWriteEventsToPublicStreams
     /// <param name="eventConverter">THe <see cref="IEventConverter" />.</param>
     /// <param name="streamWatcher">The <see cref="IStreamEventWatcher" />.</param>
     /// <param name="logger">The <see cref="ILogger" />.</param>
-    public EventsToPublicStreamsWriter(IStreams streams, IWriteEventsToStreamCollection eventsToStreamsWriter, IEventConverter eventConverter, IStreamEventWatcher streamWatcher, ILogger logger)
+    public EventsToPublicStreamsWriter(IStreams streams, IWriteEventsToStreamCollection eventsToStreamsWriter, IEventConverter eventConverter, IStreamEventWatcher streamWatcher)
```
Comment (woksin) [inline Source/Events.Store.MongoDB/EventHorizon/EventsToPublicStreamsWriter.cs:37]: This file changed

File: Source/Events.Store.MongoDB/Streams/EventsToStreamsWriter.cs [hunk 2]
```diff
@@ -42,98 +42,125 @@ public EventsToStreamsWriter(IStreams streams, IEventConverter eventConverter, I
...
-            await _streams.Get(scope, stream, cancellationToken).ConfigureAwait(false),
-            _streamFilter,
-            streamPosition =>
-                @event is CommittedExternalEvent externalEvent ?
-                    _eventConverter.ToStoreStreamEvent(externalEvent, streamPosition, partition)
-                    : _eventConverter.ToStoreStreamEvent(@event, streamPosition, partition),
-            cancellationToken).ConfigureAwait(false);
-        _streamWatcher.NotifyForEvent(scope, stream, writtenStreamPosition);
-    }
+    public Task Write(CommittedEvent @event, ScopeId scope, StreamId stream, PartitionId partition, CancellationToken cancellationToken)
```
Comment (woksin) [inline Source/Events.Store.MongoDB/Streams/EventsToStreamsWriter.cs:45]: Thisi file changed

File: Source/Events.Store.MongoDB/Streams/IWriteEventsToStreamCollection.cs [hunk 3]
```diff
@@ -20,30 +21,24 @@ public interface IWriteEventsToStreamCollection
     /// </summary>
     /// <typeparam name="TEvent">The event type.</typeparam>
     /// <param name="stream">The Stream <see cref="IMongoCollection{TDocument}" />.</param>
-    /// <param name="filter">The <see cref="FilterDefinitionBuilder{TDocument}" />.</param>
     /// <param name="createStoreEvent">A <see cref="Func{T, TResult}" /> that creates a <typeparamref name="TEvent" />.</param>
     /// <param name="cancellationToken">The <see cref="CancellationToken" />.</param>
     /// <returns>A <see cref="Task" /> representing the asynchronous operation.</returns>
     Task<StreamPosition> Write<TEvent>(
```
Comment (woksin) [inline Source/Events.Store.MongoDB/Streams/IWriteEventsToStreamCollection.cs:27]: This file changed

**Review by jakhog — APPROVED** [review]
Nice! This is a really old bug that is amazing to finally have fixed ❤️

The code is a little tricky to reason about - but with the integration tests, I feel a lot less pressure to make sure it all works. The review is now more looking at the naming and other small things - like it should be. Really nice work @woksin 👏🏻

**Pull request merged**